docs/architecture/adr-031-msg-service.md

```diff
@@ -0,0 +1,216 @@
+
+In addition to clearly specifying return types, this has the benefit of generating client and server code. On the server
+side, this is almost like an automatically generated keeper method and could maybe be used intead of keepers eventually
+(see [\#7093](https://github.com/cosmos/cosmos-sdk/issues/7093)):
+
+```go
+type MsgServer interface {
+  SubmitProposal(context.Context, *MsgSubmitProposal) (*MsgSubmitProposalResponse, error)
+}
+```
+
+On the client side, developers could take advantage of this by creating RPC implementations that encapsulate transaction
+logic. Protobuf libraries that use asynchronous callbacks, like [protobuf.js](https://github.com/protobufjs/protobuf.js#using-services)
+could use this to register callbacks for specific messages even for transactions that span include multiple `Msg`s.
+
+For backwards compatibility, existing `Msg` types should be used as the request parameter
+for `service` definitions. Newer `Msg` types which only support `service` definitions
+should use the more canonical `Msg...Request` names.
+
+### Encoding
+
+Currently, we are encoding `Msg`s as `Any` in `Tx`s which involves packing the
+binary-encoded `Msg` with its type URL.
+
+The type URL for `MsgSubmitProposal` is `/cosmos.gov.MsgSubmitProposal`. 
+
+The fully-qualified name for the `SubmitProposal` service method above is
+`/cosmos.gov.Msg/SubmitProposal` which is varies by a single `/` character.
+
+In order to encode service methods in transactions, we encode them as `Any`s in
+the same `TxBody.messages` field as other `Msg`s. We simply set `Any.type_url`
+to the full-qualified method name (ex. `/cosmos.gov.Msg/SubmitProposal`) and
+set `Any.value` to the protobuf encoding of the request message
+(`MsgSubmitProposal` in this case). 
+
+### Decoding
+
+When decoding, `TxBody.UnpackageInterfaces` will need a special case
+to detect if `Any` type URLs match the service method format (ex. `/cosmos.gov.Msg/SubmitProposal`)
+by checking for two `/` characters. Messages that are method names plus request parameters
+instead of a normal `Any` messages will get unpacked into the `ServiceMsg` struct:
+ 
+```go
+type ServiceMsg struct {
+  // MethodName is the fully-qualified service name
+  MethodName string
+  // Request is the request payload
+  Request MsgRequest
+}
+```
+
+### Routing
+
+In the future, `service` definitions may become the primary method for defining
+`Msg`s. As a starting point, we need to integrate with the SDK's existing routing
+and `Msg` interface. 
+
+To do this, `ServiceMsg` implements the `sdk.Msg` interface and its handler does the
+actual method routing, allowing this feature to be added incrementally on top of
+existing functionality.
+
+### `ServiceMsg` interface
+
+All request messages will need to implement the `MsgRequest` interface which is a
+simplified version of `Msg`, without `Route()`, `Type()` and `GetSignBytes()` which
+are no longer needed:
```

Comment:
Do the proto messages (e.g. `message MsgSubmitProposal{}`) still implement sdk.Msg? We still need `GetSignBytes()` for the legacy sign mode.

Reply:
They do as long as we are supporting amino and/or concrete `Msg` types in protobuf. If we switched to service methods everywhere or for new methods that only work with services only `MsgRequest` would be needed.